### PR TITLE
fix pycuda detaching with weakref and contex manager

### PR DESCRIPTION
**Pull request by Rodrigo-Tenorio**
Changes in #413 , although very useful, introduce a completely different coding style depending on which arguments are being given into a function.

This PR solves that by substituting `weakref.finalize` by context managers.

Must be merged after #413

### CODE REVIEW

File: tests/test_core.py [hunk 1]
```diff
@@ -496,6 +527,42 @@ def test_cumulative_twoF(self):
         self.assertTrue(reldiffmid < 0.25)
         self.assertTrue(reldiffend < 0.25)
 
+    def test_gpu_context_finalizer(self):
```
[the remark this answers is not in the document]
Comment (dbkeitel) [reply tests/test_core.py:530]: I think the `TransientGridSearch` and `TransientGridsearch.search` objects (the latter being an instance of `ComputeFstat`) will generally have the same lifetime, so either location for that test should work. But let's get #414 merged first to clean up the diff, and then see how we clean up and merge^2 this and #413 when we're next in front of the same screen.

**Comment by dbkeitel** [comment]
Ok, I've closed #413 and will rebase this one and clean it up.

**Comment by dbkeitel** [comment]
Two problems:

1. Current test implementation wants to call `search_gpu._finalizer` but that attribute is never created. I've fixed that in my local rebased branch, but not pushed yet pending discussion of 2.
2. I then extended the test to cover various combinations of manually finalizing or relying on garbage collection, and tested on a machine with GPU. Taken individually, each test runs fine, but trying to do 2 or more in one instance, it throws both "failed to create context" errors on trying to create the second search object and "invalid context" errors when trying to clean up. So it seems the user will still need to make sure that a previous context is properly cleaned up before trying to start a new one. We'll have to document that cleanly and add some extra safety checks, or go back to recommending the context manager style after all.

**Review by Rodrigo-Tenorio — COMMENTED** [review]
Latest version works after testing locally on a GPU.

**Comment by dbkeitel** [comment]
Now combined with the context manager idea from #413 to also allow for instantiating two ComputeFstat objects with pycuda usage from same session/script.

fixes https://github.com/PyFstat/PyFstat/issues/411

**Review by Rodrigo-Tenorio — COMMENTED** [review]
:+1:

**Review by dbkeitel — APPROVED** [review]
Both the extended parametrised tests and a manual "dual-use" example script working now both without a GPU available and without. Merging once readthedocs is happy.

**Pull request merged**